homeassistant/components/nice_go/coordinator.py

```diff
@@ -178,16 +184,30 @@ async def _update_refresh_token(self) -> None:
 
     async def client_listen(self) -> None:
         """Listen to the websocket for updates."""
-        self.api.event(self.on_connected)
-        self.api.event(self.on_data)
-        try:
-            await self.api.connect(reconnect=True)
-        except ApiError:
-            _LOGGER.exception("API error")
+        self._unsub_connected = self.api.listen("on_connected", self.on_connected)
+        self._unsub_data = self.api.listen("on_data", self.on_data)
+        self._unsub_connection_lost = self.api.listen(
+            "on_connection_lost", self.on_connection_lost
+        )
 
-        if not self.hass.is_stopping:
-            await asyncio.sleep(5)
-            await self.client_listen()
+        for _ in range(RECONNECT_ATTEMPTS):
+            if self.hass.is_stopping:
+                return
```

Comment:
Can't figure out how to test this line

Reply:
You can trigger `hass` stopping with `hass.bus.fire(EVENT_HOMEASSISTANT_STOP)` in a test, and then assert the background process has finished. The the backend library holding a reference to the backend process, or is it in your integration code? You could check the `Future` state.

Reply:
I pushed a potential test for it, but it's still not covered. I'm still lost :(

Reply:
Okay, I'll try to have a closer look at your code by debugging it

Reply:
I have a solution that I could push. It seems that `hass.is_stopping` is not updated directly.
Let me know if you would like me to push a commit with my solution.

Reply:
Okay, go ahead

Reply:
Done, asked @joostlek to review